### PR TITLE
[kafka] Add option to supply destination topic through context

### LINKED ISSUE #34432: Allow Kafka exporter to select topic from context
### Component(s)

exporter/kafka

### Is your feature request related to a problem? Please describe.

The kafka exporter currently allows exporting to a fixed topic or a topic from an attribute in the data.
When using an attribute, if the topic name isn't already in the data for some reason, the topic name inherently pollutes the data.

### Describe the solution you'd like

I would like to be able to pull the topic name from the context.  This would allow prior components in the pipeline to define/customize the topic name without modifying the data being exported.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tylerbenson**
**Description:** <Describe what has changed.>

Add option to get destination topic from context. This allows for upstream connectors to control the destination without polluting the data being written.

**Link to tracking Issue:** <Issue number if applicable>
Fixes #34432

**Testing:** <Describe what testing was performed and which tests were added.>
Added unit tests.

**Documentation:** <Describe the documentation added.>
Updated the component readme with the added setting.

### CODE REVIEW

File: exporter/kafkaexporter/README.md
Comment (pavolloffay) [inline exporter/kafkaexporter/README.md:29]: How is this configured in let's saay OTLP receiver? Can a user configure it?

Comment (tylerbenson) [reply exporter/kafkaexporter/README.md]: I've updated the readme and removed this line from when I was trying to use a basic string as the type. That's no longer the case.  The expectation is that users would interact with the context using the exported functions in a custom component, not via configuration.

Comment (brettbuddin) [reply exporter/kafkaexporter/README.md]: For it to be interacted with via configuration, you'd have to adopt the prefix-driven values like the attributesreceiver does, right? Things like `metadata:` for gRPC metadata, etc.

File: exporter/kafkaexporter/README.md
Comment (pavolloffay) [inline exporter/kafkaexporter/README.md:28]: `topic.WithTopic` does not seem to be a config option. 

I don't think we have to document it here, but if it is here then it would be better to document it.

Comment (tylerbenson) [reply exporter/kafkaexporter/README.md:28]: It's not a config option.
I'm not sure the right approach here.  I added it here to document the order of precedence.

Would you prefer I remove it and not document it, or add documentation in another section so it can be referenced here?

Comment (pavolloffay) [reply exporter/kafkaexporter/README.md:28]: either clarify that the `topic.WithTopic` is programmatic API or I would prefer to remove it from there and add note somewhere else.

File: .chloggen/kafka-topic-context.yaml
Comment (pavolloffay) [inline .chloggen/kafka-topic-context.yaml:25]: Please add note that this is library API change.

**Review by pavolloffay — APPROVED** [review]

File: .chloggen/kafka-topic-context.yaml
Comment (codeboten) [reply .chloggen/kafka-topic-context.yaml:25]: 👍🏻

File: exporter/kafkaexporter/topic/kafka_exporter_ctx.go
Comment (codeboten) [inline exporter/kafkaexporter/topic/kafka_exporter_ctx.go:4]: Could this be moved to `internal/kafka/topic` instead? Would that remove the need to add this to `cmd/checkapi/allowlist.txt`?

Comment (brettbuddin) [reply exporter/kafkaexporter/topic/kafka_exporter_ctx.go:4]: No, because the intent is for other components to be able to specify the topic via this opaque context key mechanism.

**Comment by pavolloffay** [comment]
@codeboten can we move this PR forward?

File: exporter/kafkaexporter/topic/kafka_exporter_ctx.go
Comment (mx-psi) [reply exporter/kafkaexporter/topic/kafka_exporter_ctx.go:4]: > No, because the intent is for other components to be able to specify the topic via this opaque context key mechanism.

Will these components be out of Collector contrib?

I don't think this should be in `exporter/kafkaexporter`, if `internal` is not the right place, then [`pkg`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg) would be the right option. Component modules are explicitly not meant to have arbitrary public API, it should just be config and the `NewFactory` function

**Review by mx-psi — DISMISSED** [review]
Sorry to block this at the last minute but I don't think we should add an allowlist exception for  this. Arbitrary Go API for building packages should be either in `internal` (if meant only for Collector contrib components) or in [`pkg`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg) (if meant also for external components, in which case I would like to have an example of which external component is going to use this)

**Comment by tylerbenson** [comment]
Oops, looks like I made a mess with the labels.  (I accidentally pushed a `make update-otel` commit.) Sorry about that.

**Comment by tylerbenson** [comment]
Thanks @crobert-1.  Sorry again.

**Comment by crobert-1** [comment]
> Thanks @crobert-1. Sorry again.

Not a problem, happens all the time! 👍

File: .chloggen/kafka-topic-context.yaml
Comment (crobert-1) [inline .chloggen/kafka-topic-context.yaml:10]:
```suggestion
note: Add option to supply destination topic through context.
```
The issue (PR in this case) number will be taken from the `issues` section below and appended to the message automatically by `.chloggen`, so we don't need to include it here manually.

File: cmd/oteltestbedcol/go.mod
Comment (crobert-1) [inline cmd/oteltestbedcol/go.mod:7]:
```suggestion
toolchain go1.22.6
```
We shouldn't be changing toolchain version here.

File: .github/CODEOWNERS
Comment (crobert-1) [inline .github/CODEOWNERS:151]: This line is added automatically from a directory having a `metadata.yaml` file that includes code owner information. There's more information [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components), and I believe you'll need to run `make generate-gh-issue-templates`, and also, if possible, update the README to have the format suggested there as well.

Comment (tylerbenson) [reply .github/CODEOWNERS]: That command didn't work for me, but I added that file and removed the line here.

Comment (mx-psi) [reply .github/CODEOWNERS]: I run the command for you on 7764e24

**Comment by mx-psi** [comment]
To make CI pass so I can merge this PR you need to run `make crosslink` and add back the component to `.github/CODEOWNERS`

**Review by MovieStoreGuy — COMMENTED** [review]
I am not sure I want to encourage passing metadata via context, in my experience it has not lead to a good time and often caused more issues than what it attempts to solve.

I also think this could create non deterministic behaviour (from a user's perspective) where data is being routed to topics that don't exist or conditions of which the data is routed to a topic do not make sense.

I would prefer users using the routing connector and having several kafka exporters that is very clearly configured and understood how data is being routed and delivered.

**Comment by MovieStoreGuy** [comment]
I am not blocking this PR because it appears there is some interest in having some form of dynamic routing configuration, but this is not the why I suggest it should done.

File: .github/CODEOWNERS
Comment (tylerbenson) [reply .github/CODEOWNERS]: Thanks!

**Comment by brettbuddin** [comment]
@MovieStoreGuy I mostly agree with you. The situation is a bit tricky for us, because we write our own custom components for Gateway Collectors that run in a multi-tenanted fashion. Due to the breadth of tenant scope it's not reasonable for everything to be explicitly declarative via configuration. In a perfect world we'd derive a topic name and print it into the telemetry attributes and leverage that value using existing component capabilities. However, we see the topic name as an internal detail and not something we want to expose to tenants via the database (post-ingestion). This is the best option we've come up with so far, because the Collector component architecture doesn't provide a mechanism for agent-local metadata propagation other than telemetry data or `context.Context`.

File: .github/CODEOWNERS
Comment (tylerbenson) [reply .github/CODEOWNERS:151]: There's some weird conflict going on.  `make crosslink` removes an extra link, but `make genoteltestbedcol` adds it back.  Any idea how to get past this?

Comment (mx-psi) [reply .github/CODEOWNERS:151]: @tylerbenson I believe you don't have to add the replace to `builder-config.yaml` from `oteltestbedcol`, since `oteltestbedcol` does not include any of the kafka components

**Review by mx-psi — APPROVED** [review]
@MovieStoreGuy Could you take a look at the comments above?

**Comment by tylerbenson** [comment]
Following up here. @pavolloffay @mx-psi @codeboten anything else I need to do?

**Comment by MovieStoreGuy** [comment]
> @MovieStoreGuy I mostly agree with you. The situation is a bit tricky for us, because we write our own custom components for Gateway Collectors that run in a multi-tenanted fashion. Due to the breadth of tenant scope and broker/topic structure it's not reasonable for everything to be explicitly declarative via configuration.

Yeah, I get where you're coming from. I also think this problem isn't strictly limited to Kafka, you still have the same problem with something like OTLP exporter as an example if you wanted to include some amount of auth/meta data. 

> In a perfect world we'd derive a topic name and print it into the telemetry attributes and leverage that value using existing component capabilities. However, we see the topic name as an internal detail and not something we want to expose to tenants via the database (post-ingestion). This is the best option we've come up with so far, because the Collector component architecture doesn't provide a mechanism for agent-local metadata propagation other than telemetry data or `context.Context`. That's a reasonable design decision to make for most situations, and if our use case is far enough out-of-bounds we can find other ways of dealing with it.

The approach does go against the recommendations of go design paradigms suggest where Context isn't recommended to be used for business logic, however, I can see why we'd need it.

I am happy to approve this, but once there is a notion of supporting tenancy, I'd like for this approach to be deprecated in favour of what is to come, does that sound reasonable to you?

**Comment by tylerbenson** [comment]
I merged from main and resolved the conflicts.
The failing test doesn't look related to my change here.

**Pull request merged**
